### PR TITLE
Switch from intelligent save to intelligent load

### LINKED ISSUE #16: Add save/load
As a follow up to https://github.com/bluesky/ophyd-epics-devices/issues/28, a method should be added to save all PV's of a device to a yaml file, and a load method should be added to set the device to a state specified in the yaml file.

### PR DESCRIPTION
**Pull request by coretl**
Fixes #16

### CODE REVIEW

**Review by olliesilvester — CHANGES_REQUESTED** [review]
This is really neat! Most of my comments are just to help with readability

File: pyproject.toml [hunk 1]
```diff
@@ -159,6 +159,9 @@ lint.select = [
     "SLF",     # self - https://docs.astral.sh/ruff/settings/#lintflake8-self
     "PLC2701", # private import - https://docs.astral.sh/ruff/rules/import-private-name/
 ]
+lint.ignore = [
```
Comment (olliesilvester) [inline pyproject.toml:162]: Change to this file might be better in a separate PR

**Comment by olliesilvester** [comment]
It would also be a good idea to test this on a real PandA before merging

File: pyproject.toml (hunk 1 above)
Comment (coretl) [reply pyproject.toml:162]: #704

File: src/ophyd_async/core/_settings.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            device = device.parent
+        return device.parent is self.device
+
+    def __setitem__(
+        self, key: SignalRW[SignalDatatypeT], value: SignalDatatypeT | None
+    ) -> None:
+        # Check the types on entry to dict to make sure we can't accidentally
+        # add a non-signal type
+        if not isinstance(key, SignalRW):
+            raise TypeError(f"Expected SignalRW, got {key}")
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_async/core/_settings.py:36]: If we ran static analysis on everything then no, but these will be used in user plans so I put a runtime check in too

File: src/ophyd_async/core/_settings.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    def __delitem__(self, key: SignalRW) -> None:
+        del self._settings[key]
+
+    def __iter__(self) -> Iterator[SignalRW]:
+        yield from iter(self._settings)
+
+    def __len__(self) -> int:
+        return len(self._settings)
+
+    def __or__(self, other: MutableMapping[SignalRW, Any]) -> Settings:
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_async/core/_settings.py:50]: I assume you mean `settings2`?
```python
>>> {1: 3} | {1: 4}
{1: 4}
```
I'll add a docstring

[the remark this answers is not in the document]
Comment (olliesilvester) [reply src/ophyd_async/core/_settings.py:50]: Yes, I've accidentally proven my point there!

File: src/ophyd_async/plan_stubs/_panda.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from bluesky.utils import MsgGenerator, plan
+
+from ophyd_async.core import Settings
+
+from ._settings import apply_settings
+
+
+@plan
+def apply_panda_settings(settings: Settings) -> MsgGenerator[None]:
+    units, others = settings.partition(lambda signal: signal.name.endswith("_units"))
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_async/plan_stubs/_panda.py:10]: Possibly, but I think I'd prefer an explicit
```python
settings = saved_settings | ignore_hdf_settings(panda) | time_seq_settings(panda.seq[1])
```

File: src/ophyd_async/plan_stubs/_settings.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+def apply_settings_if_different(
+    settings: Settings,
+    apply_plan: Callable[[Settings], MsgGenerator[None]],
+    current_settings: Settings | None = None,
+) -> MsgGenerator[None]:
+    if current_settings is None:
+        signal_values = yield from _get_values_of_signals(
+            {sig: sig for sig in settings}
+        )
+        current_settings = Settings(settings.device, signal_values)
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_async/plan_stubs/_settings.py:85]: Not quite, this is intentionally only getting the Signals in Settings, not all of them in the device. I'll add a comment to explain

**Review by olliesilvester — APPROVED** [review]
Looks good!

**Comment by coretl** [comment]
> As discussed on slack, after testing this on a real panda, it seems like `panda.children()` only gives the children for the blocks which have been type hinted.

This is resolved as discussed offline
